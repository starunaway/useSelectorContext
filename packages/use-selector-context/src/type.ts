import { Context } from 'react';

const Z = Symbol.for('ZContext');

export interface ZContext<Value> {
  [Z]: {};
  Provider: Context<Value>['Provider'];
  Consumer: null;
}

export type ZContextValue<Value = unknown> = {
  getSnapshot: () => Value;
  subscribe: (listener: Listener) => () => void;
};

export type Listener = () => void;

export type EqualityFn<T> = (a: T, b: T) => boolean;
