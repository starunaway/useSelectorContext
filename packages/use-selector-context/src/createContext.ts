import React, { ReactNode, Provider, useRef, useLayoutEffect } from 'react';
import { Listener, ZContext, ZContextValue } from './type';

function createSubscribe(listeners: Set<Listener>) {
  return function subscribe(listener: Listener) {
    listeners.add(listener);
    return function unsubscribe() {
      listeners.delete(listener);
    };
  };
}

function createProvider<Value>(originProvider: Provider<ZContextValue<Value>>) {
  const ContextProvider = ({ value, children }: { value: Value; children: ReactNode }) => {
    const contextValue = useRef<ZContextValue<Value>>();

    const realValue = useRef<Value>(value);
    const listeners = useRef(new Set<Listener>());

    if (!contextValue.current) {
      contextValue.current = {
        getSnapshot: () => realValue.current,
        subscribe: createSubscribe(listeners.current),
      };
    }

    useLayoutEffect(() => {
      realValue.current = value;
      listeners.current.forEach((listener) => listener());
    }, [value]);

    return React.createElement(originProvider, { value: contextValue.current }, children);
  };

  return ContextProvider as ZContext<Value>['Provider'];
}

export default function createContext<Value>(defaultValue: Value) {
  const contextValue: ZContextValue<Value> = {
    getSnapshot: () => defaultValue,
    subscribe: {} as any,
  };

  const context = React.createContext<ZContextValue<Value>>(contextValue);

  (context as any).Provider = createProvider(context.Provider);

  return context as unknown as ZContext<Value>;
}
