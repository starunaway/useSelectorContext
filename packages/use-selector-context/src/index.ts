export { default as useContext } from './hooks/useContext';
export { default as useGetSnapshot } from './hooks/useGetSnapshot';
export { default as createContext } from './createContext';
export * from './type';
export { default as shallowEqual } from './utils/shallowEqual';
