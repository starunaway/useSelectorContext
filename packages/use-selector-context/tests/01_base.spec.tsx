import React, { useState, StrictMode } from 'react';
import { afterEach, describe, expect, it } from 'vitest';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import { createContext, useContext } from '../src';

describe('01_basic', () => {
  afterEach(cleanup);

  it('counter', () => {
    const initialState = {
      count: 0,
      setCount: (() => {}) as React.Dispatch<React.SetStateAction<number>>,
    };
    const context = createContext(initialState);

    const Counter = () => {
      const count = useContext(context, (v) => v.count);
      const setState = useContext(context, (v) => v.setCount);
      const increment = () => setState((s) => s + 1);

      return (
        <div>
          <span data-testid="counter">{count}</span>
          <button type="button" onClick={increment}>
            +1
          </button>
        </div>
      );
    };

    const App = () => {
      const [count, setCount] = useState(0);

      return (
        <StrictMode>
          <context.Provider
            value={{
              count,
              setCount,
            }}
          >
            <Counter />
          </context.Provider>
        </StrictMode>
      );
    };

    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByTestId('counter').textContent).toEqual('1');
    expect(container).toMatchSnapshot();
  });
});
