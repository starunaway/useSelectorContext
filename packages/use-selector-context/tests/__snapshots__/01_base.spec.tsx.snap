// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`01_basic > counter 1`] = `
<div>
  <div>
    <span
      data-testid="counter"
    >
      0
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
</div>
`;

exports[`01_basic > counter 2`] = `
<div>
  <div>
    <span
      data-testid="counter"
    >
      1
    </span>
    <button
      type="button"
    >
      +1
    </button>
  </div>
</div>
`;
